src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
 }
 
 bool
-DatabaseShardImp::prepareShard(std::uint32_t shardIndex)
+DatabaseShardImp::prepareShards(std::vector<std::uint32_t> const& shardIndexes)
 {
-    auto fail = [j = j_, shardIndex](std::string const& msg) {
-        JLOG(j.error()) << "shard " << shardIndex << " " << msg;
+    auto fail = [j = j_, &shardIndexes](
+                    std::string const& msg,
+                    boost::optional<std::uint32_t> shardIndex = boost::none) {
+        std::vector<std::string> indexesAsString(shardIndexes.size());
+        std::transform(
+            shardIndexes.begin(),
+            shardIndexes.end(),
+            indexesAsString.begin(),
+            [](uint32_t const index) { return std::to_string(index); });
+
+        std::string const prequel = shardIndex
+            ? "shard " + std::to_string(*shardIndex)
+            : std::string("shard") + (shardIndexes.size() > 1 ? "s " : " ") +
+                boost::algorithm::join(indexesAsString, ", ");
+
+        JLOG(j.error()) << prequel << " " << msg;
         return false;
     };
+
     std::lock_guard lock(mutex_);
     assert(init_);
 
-    if (!canAdd_)
-        return fail("cannot be stored at this time");
+    auto historicalShardsToPrepare = 0;
 
-    if (shardIndex < earliestShardIndex())
+    for (auto const shardIndex : shardIndexes)
     {
-        return fail(
-            "comes before earliest shard index " +
-            std::to_string(earliestShardIndex()));
+        if (!canAdd_)
+            return fail("cannot be stored at this time");
+
+        if (shardIndex < earliestShardIndex())
+        {
+            return fail(
+                "comes before earliest shard index " +
+                    std::to_string(earliestShardIndex()),
+                shardIndex);
+        }
+
+        // If we are synced to the network, check if the shard index
+        // is greater or equal to the current shard.
+        auto seqCheck = [&](std::uint32_t seq) {
+            // seq will be greater than zero if valid
+            if (seq >= earliestLedgerSeq() &&
+                shardIndex >= seqToShardIndex(seq))
+                return fail("has an invalid index", shardIndex);
+            return true;
+        };
+        if (!seqCheck(app_.getLedgerMaster().getValidLedgerIndex() + 1) ||
+            !seqCheck(app_.getLedgerMaster().getCurrentLedgerIndex()))
+        {
+            return false;
+        }
+
+        if (shards_.find(shardIndex) != shards_.end())
+        {
+            JLOG(j_.debug()) << "shard " << shardIndex
```

Comment:
FWIW, not exercised in unit tests.

Reply:
Added unit tests to exercise additional paths